### ISSUE TITLE
SimCSE. 옳은 학습 방법입니까?

### ISSUE THREAD

**Issue by intrandom5** [boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13#69]
우선 제가 이해한 SimCSE의 unsupervised pretraining 방법은,
![image](https://user-images.githubusercontent.com/68216991/201005731-862a06b9-6d25-407a-85fd-0d2d6041a8d7.png)
그림으로 최대한 쉽게 그려보고 싶은데 잘 안되네요.
아무튼 Input 문장 64개(한 minibatch)가 들어오면 BERT에서 각각 다른 dropout mask가 적용된 outputs1, outputs2가 나옵니다.
그러면 outputs1과 outputs2 간의 cosine similarity를 계산하고, outputs1의 feature 하나와 가장 유사한 feature를 outputs2에서 찾아 라벨과 비교하는 방식입니다.

models/auto_model.py
>
    class SimCSE(nn.Module):
        def __init__(self, model_name, dropout_prob):
            super(SimCSE, self).__init__()
            self.bert = AutoModel.from_pretrained(
                model_name, hidden_dropout_prob=dropout_prob)
            self.cosine_similarity = nn.CosineSimilarity()
        
        def forward(self, src_ids):
            outputs1 = self.bert(src_ids).pooler_output
            outputs2 = self.bert(src_ids).pooler_output
            score = self.cosine_similarity(outputs1.unsqueeze(1), outputs2.unsqueeze(0))
        
            return outputs1, score

datasets/datasets.py
>
    class KorSTSDatasets_for_SimCSE(KorSTSDatasets):
        def __init__(self, dir, model_name, stopword=False):
            super(KorSTSDatasets_for_SimCSE, self).__init__(dir, model_name, stopword)
            self.sentences = self.s1 + self.s2
        
        def __getitem__(self, idx):
            return torch.IntTensor(self.sentences[idx])

    class Collate_fn(object):
        def __init__(self, pad_id=0, model_type="SBERT"):
            self.pad_id = pad_id
            self.model_type = model_type

        def __call__(self, batch):
            ...
            elif self.model_type == "SimCSE":
                inputs = [b for b in batch]
                labels = torch.arange(len(inputs))
                inputs = pad_sequence(inputs, batch_first=True, padding_value=self.pad_id)
                return inputs.long(), labels.long()

utils/trainer.py
>
    def step(data, model_type, device, model, criterion, outputEDA=None):
        ...
        elif model_type == "SimCSE":
            s1, label = data
            s1 = s1.to(device)
            # label = [0, 1, 2, 3, ... ,batch_size]
            label = torch.arange(s1.shape[0])
            label = label.to(device)
            logits, cos_sim = model(s1)
            loss = criterion(cos_sim, label)
            score = torch.IntTensor([0])
